src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -226,6 +229,30 @@ void CreatePrimaryRequest<I>::unlink_peer() {
   }
 
   CephContext *cct = m_image_ctx->cct;
+  if (can_remove) {
+    ldout(cct, 15) << "snap_id=" << snap_id << dendl;
+    ldout(cct, 15) << "removing snapshot with no peers "
+                   << "snap_id=" << snap_id << dendl;
+
+    cls::rbd::SnapshotNamespace snap_namespace;
+    std::string snap_name;
+
+    {
+      std::shared_lock image_locker{m_image_ctx->image_lock};
+
+      auto snap_info = m_image_ctx->get_snap_info(snap_id);
+      ceph_assert(snap_info);
+      snap_namespace = snap_info->snap_namespace;
+      snap_name = snap_info->name;
+    }
+
+    auto ctx = create_context_callback<
+      CreatePrimaryRequest<I>,
+      &CreatePrimaryRequest<I>::handle_unlink_peer>(this);
+    m_image_ctx->operations->snap_remove(snap_namespace, snap_name, ctx);
```

Comment:
Would a call to UnlinkPeerRequest with the first peer in the `m_mirror_peer_uuids` better then?

Reply:
I think if we fix that issue that the `peer` was not checked in the `info->mirror_peer_uuids` we will not need this "random" peer hack, right?

Reply:
Technically yes but if for some reasons the linked snapshot list is empty (maybe if you stop the mgr at a bad timing?) it could simply skip the snapshot and it will be never deleted.

Reply:
In this case I think I would prefer if we leaved `for (auto &peer : m_mirror_peer_uuids)` loop, and it would just go to unlink if `if (info->mirror_peer_uuids.size() == 0 || info->mirror_peer_uuids.count(peer) != 0)`.